File: app/src/main/java/com/ware/third/retrofit/RetrofitActivity.java
package com.ware.third.retrofit;

import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.ImageView;

import androidx.appcompat.app.AppCompatActivity;

import com.squareup.picasso.Picasso;
import com.ware.R;

import java.io.File;
import java.util.ArrayList;
import java.util.LinkedHashMap;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class RetrofitActivity extends AppCompatActivity implements View.OnClickListener {

    private static final String TAG = "RetrofitActivity";
    private ImageView mImageView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_retrofit);

        findViewById(R.id.bt).setOnClickListener(this);
        findViewById(R.id.bt2).setOnClickListener(this);
        mImageView = findViewById(R.id.mImageView);
        String pathname = "/storage/emulated/0/DCIM/Camera/IMG_20191206_213347.jpg";
        Picasso.get().load(new File(pathname)).fit().into(mImageView);
//
//        Picasso.Builder builder = new Picasso.Builder(WareApp.sContext);
//        builder.listener((picasso, uri, exception) -> Log.d("FeedbackP", "onBindViewHolder: ex = " + exception.getMessage())
//        );
//        builder.build().load(new File(pathname)).resize(100, 100).into(mImageView);

        LinkedHashMap<String, Info> map = new LinkedHashMap<>();
        Info value = new Info();
        value.name = "a";
        map.put("a", value);

        ArrayList<Info> strings = new ArrayList<>(map.values());

        map.get("a").name = "b";

        Log.d(TAG, "onCreate: " + strings.get(0).name);

    }

    class Info{
        String name;
    }

    @Override
    public void onClick(View v) {

        switch (v.getId()) {
            case R.id.bt:
                syncRequest();
                break;
            case R.id.bt2:

                final Retrofit retrofit = new Retrofit.Builder()
                        .baseUrl("https://api.github.com/")
                        .addConverterFactory(GsonConverterFactory.create())//Gson解析结果
                        .build();
                final GitHubService service = retrofit.create(GitHubService.class);
                Call<String> repos = service.listRepos("edreamoon");
                // 异步调用
                repos.enqueue(new Callback<String>() {

                    @Override
                    public void onResponse(Call<String> call, Response<String> response) {
                        Log.d(TAG, "onResponse: ");
                    }

                    @Override
                    public void onFailure(Call<String> call, Throwable t) {
                        Log.d(TAG, "onFailure: ");
                    }
                });

                break;
        }

    }

    private void syncRequest() {

        //同步请求
        //Response<List<Repo>> response = repos.execute();
    }
}

File: app/src/main/java/com/ware/third/retrofit/RetrofitActivity.kt
package com.ware.third.retrofit

import android.os.Bundle
import android.util.Log
import android.view.View
import com.ware.R
import com.ware.component.BaseActivity
import com.ware.http.HttpHelper
import com.ware.http.base.BaseObserver
import com.ware.http.resp.FeedArticle
import com.ware.tool.io2Main
import kotlinx.android.synthetic.main.activity_retrofit.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

private const val TAG = "RetrofitActivity"

class RetrofitActivity : BaseActivity(), View.OnClickListener {
    private val scope = MainScope()
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_retrofit)
        commonBt.setOnClickListener(this)
        rxBt.setOnClickListener(this)
        coroutineBt.setOnClickListener(this)
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.commonBt -> common()
            R.id.rxBt -> rx()
            R.id.coroutineBt -> coroutine()
        }
    }

    private fun coroutine() {
        scope.launch {
            withContext(Dispatchers.IO) {
                try {
                    val article = HttpHelper.getFeedArticleListCoroutine()
                    Log.d(TAG, "coroutine: $article")
                } catch (e: Exception) {
                    Log.d(TAG, "coroutine exception: $e")
                }

            }
        }
    }

    private fun rx() {
        addDisposable(HttpHelper.getFeedArticleListRx().compose(io2Main()).subscribeWith(object : BaseObserver<FeedArticle>() {
            override fun onSuccess(t: FeedArticle) {
                Log.d(TAG, "onSuccess: ${t.data?.size}")
            }

            override fun onFail(code: Int, msg: String?) {
                Log.d(TAG, "onFail: code = $code; msg = $msg")
            }

        }))
    }

    private fun common() {
        HttpHelper.getFeedArticleList().enqueue(object : Callback<FeedArticle> {
            override fun onFailure(call: Call<FeedArticle>, t: Throwable) {
                Log.d(TAG, "onFailure: ${t.message}")
            }

            override fun onResponse(call: Call<FeedArticle>, response: Response<FeedArticle>) {
                Log.d(TAG, "onResponse: ${response.body()}")
            }
        })
    }
}